PreCheck/doxygen-man/man3/SH_ServiceCharging.3
.TH "SH_ServiceCharging" 3 "Mardi Juillet 2 2013" "Version 0.4" "PreCheck" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SH_ServiceCharging \- 
.PP
The \fBSH_ServiceCharging\fP class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SH_ServiceCharging\&.h>\fP
.PP
Est dérivée de \fBSH_LoopingInOutStateMachine\fP\&.
.SS "Connecteurs publics"

.in +1c
.ti -1c
.RI "virtual void \fBaddChildrenReplaceTransition\fP (QAbstractState *previousState, QAbstractState *nextState)"
.br
.ti -1c
.RI "void \fBaddIOState\fP (\fBSH_InOutState\fP *state, QString field='')"
.br
.ti -1c
.RI "virtual void \fBaddState\fP (QAbstractState *state)"
.br
.ti -1c
.RI "virtual void \fBaddState\fP (\fBSH_InOutState\fP *astate, QString field='')"
.br
.ti -1c
.RI "virtual void \fBaddState\fP (\fBSH_InOutStateMachine\fP *astate, QString table='')"
.br
.ti -1c
.RI "void \fBaddStateMachine\fP (\fBSH_InOutStateMachine\fP *fsm, QString table='')"
.br
.ti -1c
.RI "void \fBsetContentValue\fP (QVariant content, QString field)"
.br
.in -1c
.SS "Signaux"

.in +1c
.ti -1c
.RI "void \fBcancelReplacement\fP ()"
.br
.ti -1c
.RI "void \fBclearAll\fP ()"
.br
.ti -1c
.RI "void \fBconfirmInput\fP ()"
.br
.ti -1c
.RI "void \fBdisplayCalendar\fP ()"
.br
.ti -1c
.RI "void \fBdisplayChoiceList\fP (QVariantList choices)"
.br
.ti -1c
.RI "void \fBdisplayFileDialog\fP ()"
.br
.ti -1c
.RI "void \fBgoNext\fP ()"
.br
.ti -1c
.RI "void \fBlimitChanged\fP ()"
.br
.ti -1c
.RI "void \fBnext\fP ()"
.br
.ti -1c
.RI "void \fBreceiveInput\fP (QString input)"
.br
.ti -1c
.RI "void \fBreplaceInput\fP (QString field)"
.br
.ti -1c
.RI "void \fBresendText\fP (QString text, bool editable=false)"
.br
.ti -1c
.RI "void \fBsendText\fP (QString text, bool editable=false)"
.br
.ti -1c
.RI "void \fBvalidateInput\fP ()"
.br
.in -1c
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBSH_ServiceCharging\fP (QString \fBname\fP, \fBQObject\fP *parent=0)"
.br
.RI "\fI\fBSH_ServiceCharging\fP\&. \fP"
.ti -1c
.RI "int \fBcurrent\fP () const "
.br
.ti -1c
.RI "QVariant \fBgetContentValue\fP (QString field)"
.br
.ti -1c
.RI "QHistoryState * \fBhistoryValue\fP (QString field)"
.br
.ti -1c
.RI "QVariantMap \fBioContent\fP () const "
.br
.ti -1c
.RI "int \fBlimit\fP () const "
.br
.ti -1c
.RI "void \fBsetCurrent\fP (int \fBcurrent\fP)"
.br
.ti -1c
.RI "void \fBsetIOcontent\fP (const QVariantMap &\fBioContent\fP)"
.br
.ti -1c
.RI "void \fBsetIOStateHistory\fP (QHistoryState *state, QString field)"
.br
.ti -1c
.RI "void \fBsetLimit\fP (int \fBlimit\fP)"
.br
.ti -1c
.RI "void \fBsetPersistentContentValue\fP (QVariant value, QString field)"
.br
.ti -1c
.RI "void \fBsetStatesNextTransition\fP (QAbstractState *previousState, QAbstractState *nextState)"
.br
.ti -1c
.RI "void \fBsetTableName\fP (const QString &\fBtableName\fP)"
.br
.ti -1c
.RI "void \fBstopLooping\fP ()"
.br
.ti -1c
.RI "QString \fBtableName\fP () const "
.br
.ti -1c
.RI "QString \fBtoString\fP ()"
.br
.in -1c
.SS "Connecteurs protégés"

.in +1c
.ti -1c
.RI "void \fBemitGoNext\fP ()"
.br
.in -1c
.SS "Fonctions membres protégées"

.in +1c
.ti -1c
.RI "QMap< QString, QHistoryState * > \fBioStatesHistory\fP () const "
.br
.ti -1c
.RI "bool \fBisERunning\fP ()"
.br
.ti -1c
.RI "void \fBonEntry\fP (QEvent *event)"
.br
.ti -1c
.RI "void \fBonExit\fP (QEvent *event)"
.br
.in -1c
.SS "Attributs protégés"

.in +1c
.ti -1c
.RI "QVariantMap \fBm_ioContent\fP"
.br
.ti -1c
.RI "QMap< QString, QHistoryState * > \fBm_ioStatesHistory\fP"
.br
.ti -1c
.RI "QString \fBm_tableName\fP"
.br
.in -1c
.SS "Propriétés"

.in +1c
.ti -1c
.RI "int \fBlimit\fP"
.br
.in -1c
.SS "Attributs privés"

.in +1c
.ti -1c
.RI "qreal \fBm_priceMin\fP"
.br
.RI "\fIm_priceMin \fP"
.ti -1c
.RI "qreal \fBm_vat\fP"
.br
.RI "\fIm_vat \fP"
.in -1c
.SH "Description détaillée"
.PP 
The \fBSH_ServiceCharging\fP class\&. 
.PP
Définition à la ligne 9 du fichier SH_ServiceCharging\&.h\&.
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "SH_ServiceCharging::SH_ServiceCharging (QStringname, \fBQObject\fP *parent = \fC0\fP)"

.PP
\fBSH_ServiceCharging\fP\&. 
.PP
\fBParamètres:\fP
.RS 4
\fIname\fP 
.br
\fIparent\fP 
.IP "\(bu" 2
\fBSH_ServiceCharging::SH_ServiceCharging\fP 
.PP
.RE
.PP

.PP
Définition à la ligne 13 du fichier SH_ServiceCharging\&.cpp\&.
.PP
Références SH_InOutStateMachine::addState(), SH_QuestionState::answerInvalid(), SH_QuestionState::answerValid(), SH_InOutStateMachine::confirmInput(), SH_DatabaseManager::execSelectQuery(), SH_DatabaseManager::getInstance(), SH_GenericState::goNext(), m_priceMin, m_vat, SH_NamedObject::name(), SH_DatabaseContentQuestionState::rawInput(), SH_InOutStateMachine::receiveInput(), SH_QuestionState::setInput(), SH_InOutState::setInput(), SH_DatabaseContentQuestionState::setOutput(), SH_InOutState::setOutput(), SH_LoopingInOutStateMachine::setStatesNextTransition(), SH_InOutState::setVisibility(), SH_LoopingInOutStateMachine::stopLooping(), et SH_InOutStateMachine::validateInput()\&.
.PP
.nf
13                                                                     :
14     SH_LoopingInOutStateMachine("CHARGEDSERVICES",name, 0, parent), m_priceMin(0\&.0)
15 {
16     SH_DatabaseContentQuestionState* service = new SH_DatabaseContentQuestionState("Veuillez sélectionner une prestation ou appuyer sur la touche \"VALIDER\" pour cesser d'ajouter des prestations", "choose service in service charging","SERVICES","CODE");
17     SH_InOutState*serviceId = new SH_InOutState("","service id in service charging");
18     serviceId->setVisibility(false);
19     SH_StringQuestionState* serviceName = new SH_StringQuestionState("Veuillez entrer ce qui sera affiché sur la facture", "service name in service charging",1);
20     SH_DecimalQuestionState* price = new SH_DecimalQuestionState("", "price in service charging",-Q_INFINITY,Q_INFINITY);
21     SH_DecimalQuestionState* quantity = new SH_DecimalQuestionState("", "quantity in service charging",1);
22     SH_DatabaseContentQuestionState* vat = new SH_DatabaseContentQuestionState("", "vat in service charging","TAXES","PERCENTAGE","ENABLED='1'");
23     QFinalState* final = new QFinalState();
24     connect(service, &SH_QuestionState::answerInvalid, [=]() {
25         int in = service->rawInput()\&.toInt();
26         if(in == -1 || in == 0) {
27             emit service->goNext();
28         }
29     });
30     connect(service, &SH_QuestionState::answerValid, [=]() {
31         if(service->rawInput()\&.toInt() > -1) {
32             QString name;
33             QStringList list;
34             list\&.append("PRINTEDNAME");
35             list\&.append("PRICEMIN");
36             list\&.append("PRICEMAX");
37             list\&.append("VAT_PERCENTAGE");
38             list\&.append("ID");
39             QSqlQuery result = SH_DatabaseManager::getInstance()->execSelectQuery("SERVICESINFOS", list, QString("CODE=%1")\&.arg(service->rawInput()\&.toString()));
40             result\&.next();
41             QSqlRecord record = result\&.record();
42             name= record\&.value(0)\&.toString();
43             m_priceMin =record\&.value(1)\&.toDouble();
44             m_vat =record\&.value(3)\&.toDouble();
45             serviceId->setInput(record\&.value(4)\&.toInt());
46             serviceName->setInput(name);
47             price->setOutput(QString("Le prix proposé pour cette prestation est : %1\&. Son prix minimum est %1 et son prix maximum %2\&.\nVeuillez entrer un nouveau prix ou appuyer sur la touche \"CONFIRMER\"")\&.arg(record\&.value(1)\&.toString())\&.arg(record\&.value(2)\&.toString()));
48             vat->setOutput(QString("Cette prestation est associée à une TVA de %1\%\&.\nVeuillez entrer une autre TVA à appliquer ou appuyer sur la touche \"CONFIRMER\"")\&.arg(record\&.value(3)\&.toString()));
49             serviceName->setVisibility(false);
50         }
51     });
52     connect(quantity, &QState::entered, [=]() {
53         connect(this, &SH_InOutStateMachine::receiveInput, [=](QString in) {
54             QString newInput;
55             if(in\&.right(in\&.length() - 1)\&.toInt() != 0) {
56                 newInput = in\&.right(in\&.length() - 1);
57             }
58             emit receiveInput(newInput);
59         });
60     });
61     connect(price, &QState::entered, [=]() {
62         connect(this, &SH_InOutStateMachine::confirmInput, [=]() {
63             price->setInput(m_priceMin);
64         });
65     });
66     connect(vat, &QState::entered, [=]() {
67         connect(this, &SH_InOutStateMachine::confirmInput, [=]() {
68             vat->setInput(m_vat);
69         });
70     });
71     this->addState(final);
72     this->addState(service, "");
73     this->addState(serviceId, "SERVICE_ID");
74     this->addState(serviceName, "PRINTEDNAME");
75     this->addState(price, "CHARGEDUNITPRICE");
76     this->addState(quantity, "QUANTITY");
77     this->addState(vat, "CHARGEDVAT");
78     this->setStatesNextTransition(service, serviceId);
79     this->setStatesNextTransition(serviceId, serviceName);
80     this->setStatesNextTransition(serviceName, quantity);
81     this->setStatesNextTransition(quantity, price);
82     this->setStatesNextTransition(price, vat);
83     this->setStatesNextTransition(vat, final);
84     this->setInitialState(service);
85     connect(this, &SH_InOutStateMachine::validateInput, this, &SH_LoopingInOutStateMachine::stopLooping);
86 }
.fi
.SH "Documentation des fonctions membres"
.PP 
.SS "SH_InOutStateMachine::addChildrenReplaceTransition (QAbstractState *previousState, QAbstractState *nextState)\fC [virtual]\fP, \fC [slot]\fP, \fC [inherited]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIpreviousState\fP 
.br
\fInextState\fP 
.RE
.PP

.PP
Définition à la ligne 227 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::historyValue(), SH_GenericStateMachine::next(), et SH_InOutStateMachine::replaceInput()\&.
.PP
Référencé par SH_LoopingInOutStateMachine::setStatesNextTransition()\&.
.PP
.nf
228 {
229     SH_GenericState* genPreviousState = qobject_cast<SH_GenericState*>(previousState);
230     if(genPreviousState) {
231         /*connect(this, &SH_GenericStateMachine::entered, [=]() {*/
232         /*à faire au moment de l'entrée dans l'état previousState*/
233         /*connect(genPreviousState, &QAbstractState::entered, [=]() {*/
234         connect(this, &SH_InOutStateMachine::replaceInput, [=](QString field) {
235             /*après avoir demandé à revenir sur un état précédent, on attend la fin de l'état actuel puis on retourne à l'historique de l'état désiré; celui-ci fini, on passe à l'état qui aurait du suivre celui pendant lequel on a demandé à revenir sur un état précédent*/
236             QHistoryState* hState = historyValue(field);
237             if(hState) {
238                 /*si l'historique existe (on a déjà quitté l'état voulu)*/
239                 hState->parentState()->addTransition(hState->parentState(), SIGNAL(next()), nextState);
240                 genPreviousState->addTransition(genPreviousState, SIGNAL(next()), hState);
241             }
242         });
243         /*});*/
244         /*});*/
245     }
246 }
.fi
.SS "SH_InOutStateMachine::addIOState (\fBSH_InOutState\fP *state, QStringfield = \fC''\fP)\fC [slot]\fP, \fC [inherited]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIstate\fP 
.br
\fIfield\fP 
.RE
.PP

.PP
Définition à la ligne 84 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::addState(), SH_ConfirmationState::confirmInput(), SH_MessageManager::debugMessage(), SH_InOutStateMachine::displayCalendar(), SH_DatabaseContentQuestionState::displayChoiceList(), SH_InOutStateMachine::displayChoiceList(), SH_InOutStateMachine::displayFileDialog(), SH_MessageManager::infoMessage(), SH_InOutState::rawInput(), SH_InOutStateMachine::receiveInput(), SH_InOutState::resendInput(), SH_InOutStateMachine::resendText(), SH_InOutState::sendOutput(), SH_InOutStateMachine::sendText(), SH_InOutStateMachine::setContentValue(), SH_InOutState::setInput(), SH_InOutStateMachine::setIOStateHistory(), et SH_InOutStateMachine::validateInput()\&.
.PP
Référencé par SH_InOutStateMachine::addState()\&.
.PP
.nf
85 {
86     SH_InOutState* state = qobject_cast<SH_InOutState *>(astate);
87     if(state) {
88         /*connect(this, &SH_GenericStateMachine::entered, [=]() {*/
89         /*à faire au moment de l'entrée dans l'état state*/
90         /*connect(state, &QAbstractState::entered, [=]() {*/
91         /* la réception d'une valeur entraîne son enregistrement comme entrée de l'utilisateur auprès de l'état*/
92         connect(state, &SH_InOutState::sendOutput, [=](QVariant out) { if(out\&.isValid()) {
93                 SH_MessageManager::infoMessage(out\&.toString(),"reçu de l'état");
94                 emit this->sendText(out\&.toString(), false);
95             }});
96         connect(this, &SH_InOutStateMachine::receiveInput, state, &SH_InOutState::setInput, Qt::QueuedConnection);
97         connect(state, &SH_InOutState::resendInput, [=](QVariant in) {  if(in\&.isValid()) {
98                 SH_MessageManager::infoMessage(in\&.toString(),"envoyé par l'état");
99                 emit this->resendText(in\&.toString(), true);
100             }});
101         SH_MessageManager::debugMessage("salioute");
102         SH_ValidationState *validationState = qobject_cast<SH_ValidationState*>(astate);
103         if(validationState) {
104             connect(this, &SH_InOutStateMachine::validateInput, validationState, &SH_ValidationState::confirmInput, Qt::QueuedConnection);
105         }
106 
107         SH_ConfirmationState *confirmationState = qobject_cast<SH_ConfirmationState*>(astate);
108         if(confirmationState) {
109             connect(this, &SH_InOutStateMachine::validateInput, confirmationState, &SH_ConfirmationState::confirmInput, Qt::QueuedConnection);
110         }
111 
112         SH_DateQuestionState *dateState = qobject_cast<SH_DateQuestionState*>(astate);
113         if(dateState) {
114             emit this->displayCalendar();
115         }
116 
117         SH_DatabaseContentQuestionState *choiceState = qobject_cast<SH_DatabaseContentQuestionState*>(astate);
118         if(choiceState) {
119             connect(this, &SH_InOutStateMachine::displayChoiceList, choiceState, &SH_DatabaseContentQuestionState::displayChoiceList, Qt::QueuedConnection);
120         }
121 
122         SH_FileSelectionState *fileState = qobject_cast<SH_FileSelectionState*>(astate);
123         if(fileState) {
124             emit this->displayFileDialog();
125         }
126         SH_MessageManager::debugMessage("salioute bis");
127         /*});*/
128         connect(state, &QAbstractState::exited, [=]() {
129             if(!field\&.isEmpty()) {
130                 setContentValue(state->rawInput(), field);
131 
132                 /*gestion de l'historique des états pour pouvoir revenir à l'état state après l'avoir quitté*/
133                 QHistoryState* hState = new QHistoryState(state);
134                 setIOStateHistory(hState, field);
135             }
136             /*plus aucune action sur l'état ne pourra être provoquée par la machine*/
137             state->disconnect(this);
138         });
139         /*});*/
140     }
141     SH_MessageManager::debugMessage("salioute bis bis");
142     QAbstractState* abstate = qobject_cast<QAbstractState *>(astate);
143     if(abstate) {
144         SH_InOutStateMachine::addState(abstate);
145     }
146 }
.fi
.SS "void SH_InOutStateMachine::addState (QAbstractState *state)\fC [virtual]\fP, \fC [slot]\fP, \fC [inherited]\fP"

.PP
Définition à la ligne 68 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_GenericStateMachine::addState(), et SH_MessageManager::debugMessage()\&.
.PP
Référencé par SH_InOutStateMachine::addIOState(), SH_InOutStateMachine::addStateMachine(), SH_BillingCreationStateMachine::SH_BillingCreationStateMachine(), SH_ClientCreationStateMachine::SH_ClientCreationStateMachine(), et SH_ServiceCharging()\&.
.PP
.nf
69 {
70     SH_MessageManager::debugMessage("chalut");
71     SH_GenericStateMachine::addState(state);
72 }
.fi
.SS "void SH_InOutStateMachine::addState (\fBSH_InOutState\fP *astate, QStringfield = \fC''\fP)\fC [virtual]\fP, \fC [slot]\fP, \fC [inherited]\fP"

.PP
Définition à la ligne 74 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::addIOState(), et SH_MessageManager::debugMessage()\&.
.PP
.nf
75 {
76     SH_MessageManager::debugMessage("salioute state");
77     SH_InOutStateMachine::addIOState(state, field);
78 }
.fi
.SS "void SH_InOutStateMachine::addState (\fBSH_InOutStateMachine\fP *astate, QStringtable = \fC''\fP)\fC [virtual]\fP, \fC [slot]\fP, \fC [inherited]\fP"

.PP
Définition à la ligne 152 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::addStateMachine()\&.
.PP
.nf
153 {
154     SH_InOutStateMachine::addStateMachine(fsm, table);
155 }
.fi
.SS "SH_InOutStateMachine::addStateMachine (\fBSH_InOutStateMachine\fP *fsm, QStringtable = \fC''\fP)\fC [slot]\fP, \fC [inherited]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIfsm\fP 
.RE
.PP

.PP
Définition à la ligne 157 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::addState(), SH_InOutStateMachine::cancelReplacement(), SH_InOutStateMachine::confirmInput(), SH_MessageManager::debugMessage(), SH_InOutStateMachine::displayCalendar(), SH_InOutStateMachine::displayChoiceList(), SH_InOutStateMachine::receiveInput(), SH_InOutStateMachine::replaceInput(), SH_InOutStateMachine::resendText(), SH_InOutStateMachine::sendText(), SH_InOutStateMachine::setTableName(), et SH_InOutStateMachine::validateInput()\&.
.PP
Référencé par SH_InOutStateMachine::addState()\&.
.PP
.nf
157                                                                                       {
158     SH_MessageManager::debugMessage("salioute machine");
159     SH_InOutStateMachine* fsm = qobject_cast<SH_InOutStateMachine *>(astate);
160     if(fsm) {
161         if(!table\&.isEmpty()) {
162             fsm->setTableName(table);
163         }
164         /*connect(this, &SH_InOutStateMachine::entered, [=]() {*/
165         /*à faire au moment de l'entrée dans la machine d'état fsm*/
166         /*connect(fsm, &SH_InOutStateMachine::entered, [=]() {*/
167         connect(this, &SH_InOutStateMachine::receiveInput, fsm, &SH_InOutStateMachine::receiveInput,Qt::QueuedConnection);
168         connect(this, &SH_InOutStateMachine::sendText, fsm, &SH_InOutStateMachine::sendText,Qt::QueuedConnection);
169         connect(this, &SH_InOutStateMachine::resendText, fsm, &SH_InOutStateMachine::resendText,Qt::QueuedConnection);
170         connect(this, &SH_InOutStateMachine::confirmInput, fsm, &SH_InOutStateMachine::confirmInput,Qt::QueuedConnection);
171         connect(this, &SH_InOutStateMachine::validateInput, fsm, &SH_InOutStateMachine::validateInput,Qt::QueuedConnection);
172         connect(this, &SH_InOutStateMachine::replaceInput, fsm, &SH_InOutStateMachine::replaceInput,Qt::QueuedConnection);
173         connect(this, &SH_InOutStateMachine::cancelReplacement, fsm, &SH_InOutStateMachine::cancelReplacement,Qt::QueuedConnection);
174         connect(this, &SH_InOutStateMachine::displayCalendar, fsm, &SH_InOutStateMachine::displayCalendar,Qt::QueuedConnection);
175         connect(this, &SH_InOutStateMachine::displayChoiceList, fsm, &SH_InOutStateMachine::displayChoiceList,Qt::QueuedConnection);
176         /* });*/
177 
178         /*à faire au moment de la sortie de la machine d'état fsm*/
179         /*connect(fsm, &SH_InOutStateMachine::exited, [=]() {*/
180         /*plus aucune action sur la machine d'état fille ne pourra être provoquée par la machine mère*/
181         /*fsm->disconnect(this);*/
182         /*});*/
183         /*});*/
184     }
185     QAbstractState* abstate = qobject_cast<QAbstractState *>(astate);
186     if(abstate) {
187         SH_InOutStateMachine::addState(abstate);
188     }
189 }
.fi
.SS "SH_InOutStateMachine::cancelReplacement ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_InOutStateMachine::addStateMachine(), et SH_ApplicationCore::cancelReplacement()\&.
.SS "SH_InOutStateMachine::clearAll ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_ApplicationCore::launchStateMachine(), et SH_InOutStateMachine::setStatesNextTransition()\&.
.SS "SH_InOutStateMachine::confirmInput ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_InOutStateMachine::addStateMachine(), SH_ApplicationCore::receiveConfirmation(), SH_BillingCreationStateMachine::SH_BillingCreationStateMachine(), et SH_ServiceCharging()\&.
.SS "SH_LoopingInOutStateMachine::current () const\fC [inherited]\fP"

.PP
\fBRenvoie:\fP
.RS 4
int 
.RE
.PP

.PP
Définition à la ligne 17 du fichier SH_LoopingIOStateMachine\&.cpp\&.
.PP
Références SH_LoopingInOutStateMachine::m_current\&.
.PP
Référencé par SH_LoopingInOutStateMachine::setCurrent(), et SH_BillingCreationStateMachine::SH_BillingCreationStateMachine()\&.
.PP
.nf
18 {
19     return m_current;
20 }
.fi
.SS "SH_InOutStateMachine::displayCalendar ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_InOutStateMachine::addIOState(), SH_InOutStateMachine::addStateMachine(), et SH_ApplicationCore::launchStateMachine()\&.
.SS "void SH_InOutStateMachine::displayChoiceList (QVariantListchoices)\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_InOutStateMachine::addIOState(), SH_InOutStateMachine::addStateMachine(), et SH_ApplicationCore::launchStateMachine()\&.
.SS "SH_InOutStateMachine::displayFileDialog ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_InOutStateMachine::addIOState(), et SH_ApplicationCore::launchStateMachine()\&.
.SS "void SH_GenericStateMachine::emitGoNext ()\fC [protected]\fP, \fC [slot]\fP, \fC [inherited]\fP"

.PP
Définition à la ligne 52 du fichier SH_GenericDebugableStateMachine\&.cpp\&.
.PP
Références SH_GenericStateMachine::next()\&.
.PP
Référencé par SH_GenericStateMachine::SH_GenericStateMachine()\&.
.PP
.nf
53 {
54     if(isRunning()) {
55         emit next();
56     }
57 }
.fi
.SS "SH_InOutStateMachine::getContentValue (QStringfield)\fC [inherited]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIfield\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
QVariant 
.RE
.PP

.PP
Définition à la ligne 39 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::m_ioContent\&.
.PP
Référencé par SH_ApplicationCore::billOpened(), et SH_BillingCreationStateMachine::SH_BillingCreationStateMachine()\&.
.PP
.nf
40 {
41     return m_ioContent\&.value(field);
42 }
.fi
.SS "SH_GenericStateMachine::goNext ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_AddressCreationStateMachine::SH_AddressCreationStateMachine(), et SH_GenericStateMachine::SH_GenericStateMachine()\&.
.SS "SH_InOutStateMachine::historyValue (QStringfield)\fC [inherited]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIfield\fP 
.RE
.PP
\fBRenvoie:\fP
.RS 4
QHistoryState 
.RE
.PP

.PP
Définition à la ligne 219 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::m_ioStatesHistory\&.
.PP
Référencé par SH_InOutStateMachine::addChildrenReplaceTransition(), et SH_LoopingInOutStateMachine::setStatesNextTransition()\&.
.PP
.nf
220 {
221     return m_ioStatesHistory\&.value(field);
222 }
.fi
.SS "SH_InOutStateMachine::ioContent () const\fC [inherited]\fP"

.PP
\fBRenvoie:\fP
.RS 4
QVariantMap 
.RE
.PP

.PP
Définition à la ligne 23 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::m_ioContent\&.
.PP
Référencé par SH_InOutStateMachine::setIOcontent()\&.
.PP
.nf
24 {
25     return m_ioContent;
26 }
.fi
.SS "SH_InOutStateMachine::ioStatesHistory () const\fC [protected]\fP, \fC [inherited]\fP"

.PP
\fBRenvoie:\fP
.RS 4
QMap<QString, QHistoryState *> 
.RE
.PP

.PP
Définition à la ligne 194 du fichier SH_IOStateMachine\&.cpp\&.
.PP
Références SH_InOutStateMachine::m_ioStatesHistory\&.
.PP
Référencé par SH_InOutStateMachine::setIOStatesHistory()\&.
.PP
.nf
195 {
196     return m_ioStatesHistory;
197 }
.fi
.SS "SH_InOutStateMachine::isERunning ()\fC [protected]\fP, \fC [inherited]\fP"

.PP
\fBRenvoie:\fP
.RS 4
bool 
.RE
.PP

.SS "int SH_LoopingInOutStateMachine::limit () const\fC [inherited]\fP"

.PP
Référencé par SH_LoopingInOutStateMachine::setLimit()\&.
.SS "SH_LoopingInOutStateMachine::limitChanged ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_LoopingInOutStateMachine::setLimit()\&.
.SS "SH_GenericStateMachine::next ()\fC [signal]\fP, \fC [inherited]\fP"

.PP
Référencé par SH_InOutStateMachine::addChildrenReplaceTransition(), SH_GenericStateMachine::emitGoNext(), SH_GenericStateMachine::setStatesNextTransition(), et SH_LoopingInOutStateMachine::setStatesNextTransition()\&.
.SS "SH_GenericStateMachine::onEntry (QEvent *event)\fC [protected]\fP, \fC [inherited]\fP"

.PP
\fBParamètres:\fP
.RS 4
\fIevent\fP 
.RE
.PP

.PP

src/PreCheck/logic/SH_ServiceCharging.cpp
#include "SH_ServiceCharging.h"
#include "SH_ValidationState.h"
#include "SH_DatabaseContentQuestionState.h"
#include "SH_StringQuestionState.h"
#include "SH_DecimalQuestionState.h"
#include "SH_DatabaseManager.h"
#include <QtSql>
/*namespace SimplHotel
{*/
/*!
 * * \details \~french SH_ServiceCharging::SH_ServiceCharging
 */
SH_ServiceCharging::SH_ServiceCharging(QString name, QObject *parent) :
    SH_LoopingInOutStateMachine("CHARGEDSERVICES",name, 0, parent), m_priceMin(0.0)
{
    SH_DatabaseContentQuestionState* service = new SH_DatabaseContentQuestionState("Veuillez sélectionner une prestation ou appuyer sur la touche \"VALIDER\" pour cesser d'ajouter des prestations", "choose service in service charging","SERVICES","CODE");
    SH_InOutState*serviceId = new SH_InOutState("","service id in service charging");
    serviceId->setVisibility(false);
    SH_StringQuestionState* serviceName = new SH_StringQuestionState("Veuillez entrer ce qui sera affiché sur la facture", "service name in service charging",1);
    SH_DecimalQuestionState* price = new SH_DecimalQuestionState("", "price in service charging",-Q_INFINITY,Q_INFINITY);
    SH_DecimalQuestionState* quantity = new SH_DecimalQuestionState("", "quantity in service charging",1);
    SH_DatabaseContentQuestionState* vat = new SH_DatabaseContentQuestionState("", "vat in service charging","TAXES","PERCENTAGE","ENABLED='1'", true);
    QFinalState* final = new QFinalState();
    connect(service, &SH_QuestionState::answerInvalid, [=]() {
        int in = service->rawInput().toInt();
        if(in == -1 || in == 0) {
            emit service->goNext();
        }
    });
    connect(service, &SH_QuestionState::answerValid, [=]() {
        if(service->rawInput().toInt() > -1) {
            QString name;
            QStringList list;
            list.append("PRINTEDNAME");
            list.append("PRICEMIN");
            list.append("PRICEMAX");
            list.append("VAT_PERCENTAGE");
            list.append("ID");
            QSqlQuery result = SH_DatabaseManager::getInstance()->execSelectQuery("SERVICESINFOS", list, QString("CODE=%1").arg(service->rawInput().toString()));
            result.next();
            QSqlRecord record = result.record();
            name= record.value(0).toString();
            m_priceMin =record.value(1).toDouble();
            m_vat =record.value(3).toDouble();
            serviceId->setInput(record.value(4).toInt());
            serviceName->setInput(name);
            price->setOutput(QString("Le prix proposé pour cette prestation est : %1. Son prix minimum est %1 et son prix maximum %2.\nVeuillez entrer un nouveau prix ou appuyer sur la touche \"CONFIRMER\"").arg(record.value(1).toString()).arg(record.value(2).toString()));
            vat->setOutput(QString("Cette prestation est associée à une TVA de %1\%.\nVeuillez entrer une autre TVA à appliquer ou appuyer sur la touche \"CONFIRMER\"").arg(record.value(3).toString()));
            serviceName->setVisibility(false);
        }
    });
    connect(quantity, &QState::entered, [=]() {
        connect(this, &SH_InOutStateMachine::receiveInput, [=](QString in) {
            QString newInput;
            if(in.right(in.length() - 1).toInt() != 0) {
                newInput = in.right(in.length() - 1);
            }
            emit receiveInput(newInput);
        });
    });
    connect(price, &QState::entered, [=]() {
        connect(this, &SH_InOutStateMachine::confirmInput, [=]() {
            price->setInput(m_priceMin);
        });
    });
    connect(vat, &QState::entered, [=]() {
        connect(this, &SH_InOutStateMachine::confirmInput, [=]() {
            vat->setInput(m_vat);
        });
    });
    this->addState(final);
    this->addState(service, "");
    this->addState(serviceId, "SERVICE_ID");
    this->addState(serviceName, "PRINTEDNAME");
    this->addState(price, "CHARGEDUNITPRICE");
    this->addState(quantity, "QUANTITY");
    this->addState(vat, "CHARGEDVAT");
    this->setStatesNextTransition(service, serviceId);
    this->setStatesNextTransition(serviceId, serviceName);
    this->setStatesNextTransition(serviceName, quantity);
    this->setStatesNextTransition(quantity, price);
    this->setStatesNextTransition(price, vat);
    this->setStatesNextTransition(vat, final);
    this->setInitialState(service);
    connect(this, &SH_InOutStateMachine::validateInput, this, &SH_LoopingInOutStateMachine::stopLooping);
}
/*}*/
